src/compose.ts

```diff
@@ -346,3 +395,141 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+function normalizeAndValidateFilePath(envFile: StringOrList): StringOrList {
+	// use a set to store only unique normalized file paths
+	const intermediateEnvfile: Set<string> = new Set();
+	if (!Array.isArray(envFile)) {
+		envFile = [envFile];
+	}
+	for (let envFilePath of envFile) {
+		envFilePath = path.normalize(envFilePath);
+		if (path.isAbsolute(envFilePath)) {
+			throw new ValidationError(
+				`Absolute filepath not allowed: ${envFilePath}`,
+			);
+		}
+		if (envFilePath.startsWith('..')) {
+			throw new ValidationError(
+				`Directory traversing not allowed : ${envFilePath}`,
+			);
+		}
+		if (envFilePath.includes('*')) {
+			throw new ValidationError(`Wildcards not allowed : ${envFilePath}`);
+		}
+		intermediateEnvfile.add(envFilePath);
+	}
+	// spread set and return as array
+	return [...intermediateEnvfile];
+}
+
+async function expandContext(
+	normalizedComposition: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	// read all env_file delcared file paths from the composition
+	// normalize the environment variables
+	const expandedEnvironmentFiles: Dict<Dict<string>> =
+		await readAndEnvFilesFromComposition(normalizedComposition, fileResolverCb);
+	// assign all normalized envrionment variables to the services in the composition
+	assignExpandedEnvFilesToComposition(
+		normalizedComposition,
+		expandedEnvironmentFiles,
+	);
+	return normalizedComposition;
+}
+
+async function readAndEnvFilesFromComposition(
+	normalizedComposition: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Dict<Dict<string>>> {
+	const normalizedEnvFileVariables: Dict<Dict<string>> = {};
+	for (const service of Object.values(normalizedComposition.services)) {
+		let envFilePaths = service.env_file;
+		if (!!envFilePaths) {
+			if (!Array.isArray(envFilePaths)) {
+				envFilePaths = [envFilePaths];
+			}
+			for (const envFilePath of envFilePaths) {
+				if (!(envFilePath in normalizedEnvFileVariables)) {
+					normalizedEnvFileVariables[envFilePath] = await readAndExpandEnvFile(
+						envFilePath,
+						fileResolverCb,
+					);
+				}
+			}
+		}
+	}
+	return normalizedEnvFileVariables;
+}
+
+function assignExpandedEnvFilesToComposition(
+	normalizedComposition: Composition,
```

Comment:
redundant to add the `normalized` prefix to everything that has been normalized

Reply:
agreed, deleted preefix